src/test/java/org/mskcc/cbio/oncokb/web/rest/UsageAnalysisControllerIT.java

```diff
@@ -0,0 +1,931 @@
+    private void safeSetNestedValueInFilesObject(
+      float value,
+      String filePath,
+      int index,
+      String... keys
+    ) {
+      KeyValuePair<String, JsonObject> pair = fetchJsonObjectFromFilesObject(
+        filePath,
+        index,
+        keys
+      );
+      String lastKey = pair.key;
+      JsonObject obj = pair.value;
+      obj.addProperty(lastKey, value);
+    }
+
+    /**
+     * Set an empty object to the value found in the files JSON object based on the JSON path created by the
+     * keys. If an object is missing in the JSON path then it's created.
+     * @param value The value that is going to be added to value located in the JSON path.
+     * @param filePath First key in the files object.
+     * @param index Array index of the array the {@code filePath} parameter points to.
+     * @param keys The remaining JSON path keys.
+     */
+    private void safeSetNestedEmptyObjectInFilesObject(String... keys) {
+      KeyValuePair<String, JsonObject> pair = fetchJsonObjectFromFilesObject(
+        keys
+      );
+      String lastKey = pair.key;
+      JsonObject obj = pair.value;
+      if (!obj.has(lastKey)) {
+        obj.add(lastKey, new JsonObject());
+      }
+    }
+
+    private void addUsers(String[] userNames) throws Exception {
+      LocalDate today = TimeUtil.getCurrentNYTime().toLocalDate();
+      int currentYear = today.getYear();
+      int userId = -1;
+      for (String userName : userNames) {
+        int totalUsage = 0;
+        userId++;
+        // force ordering to consistent since we are doing strict checks on
+        // asserts for json
+        userName = userId + userName;
+        HashMap<String, Integer> userResourceUsage = new HashMap<>();
+        for (String oncokbEndpoint : sampleOncokbEndpoints) {
+          userResourceUsage.put(oncokbEndpoint, 0);
+        }
+
+        int companyIndex = userId % fakeCompanies.length;
+        String companyName = fakeCompanies[companyIndex];
+
+        User user = createMockUser(userId, userName, companyName);
+        users.add(user);
+
+        CompanyDTO companyDTO = createMockCompany(companyIndex, companyName);
+
+        UserDTO userDto = createMockUserDto(user, companyDTO);
+        userDtos.add(userDto);
+
+        String usageUserEndpoint = "/api/usage/users/" + user.getId();
+        String[] userUsageSummaryEndpoints = {
+          "/api/usage/summary/users",
+          "/api/usage/summary/users?companyId=" + companyDTO.getId(),
+        };
+
+        for (int yearIndex = 0; yearIndex < 4; yearIndex++) {
+          int year = currentYear - yearIndex;
+          Boolean isThisYear = yearIndex == 0;
+
+          DateTimeFormatter dayFormatter = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd"
+          );
+          DateTimeFormatter monthFormatter = DateTimeFormatter.ofPattern(
+            "yyyy-MM"
+          );
+
+          String yearKey = String.valueOf(year);
+
+          for (int month = 1; month <= 12; month++) {
```

Comment:
Interesting testing stratagy. I'm not testing expert. But would dynamic mock data introduces unpredictability?

Reply:
Yes, but I'm hard coding the seed for the random number generator so the data is always the same unless the calls to the random number change which would only happen if we make changes to the mock data.

I thought about making just a bunch of JSON files, but I felt like this would be easier to make changes to.

That said, I could switch this to reading a bunch of JSON files instead of dynamically generating the mock data if you feel it's better to maintain.

Reply:
The data is not always the same because you also use `today` as timestamp. The file will be different every day. Is that ok given the use case here? If it's ok, fine to me to keep the current logic.

Reply:
I forgot about that part, but since the endpoint logic is dependent on today's date there isn't a way I can think of that wouldn't require changes to `UsageAnalysisController`. You'd have to mock out all library calls to `java.time`.

I don't think it's worth mocking that logic since this will only shift the data daily.

Reply:
I figured it out. I can use the built-in java clock object.